pkg/frontend/v2/frontend_scheduler_adapter.go

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package v2
+
+import (
+	"context"
+	"time"
+
+	"github.com/grafana/dskit/httpgrpc"
+	"github.com/grafana/dskit/tenant"
+
+	"github.com/grafana/mimir/pkg/frontend/querymiddleware"
+	"github.com/grafana/mimir/pkg/querier"
+	"github.com/grafana/mimir/pkg/scheduler/schedulerpb"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+type frontendToSchedulerAdapter struct {
+	cfg    Config
+	limits Limits
+}
+
+func (a *frontendToSchedulerAdapter) frontendToSchedulerEnqueueRequest(
+	req *frontendRequest, frontendAddr string,
+) (*schedulerpb.FrontendToScheduler, error) {
+	addlQueueDims, err := a.extractAdditionalQueueDimensions(req.ctx, req.request, time.Now())
+	if err != nil {
+		return nil, err
+	}
+	return &schedulerpb.FrontendToScheduler{
+		Type:                      schedulerpb.ENQUEUE,
+		QueryID:                   req.queryID,
+		UserID:                    req.userID,
+		HttpRequest:               req.request,
+		FrontendAddress:           frontendAddr,
+		StatsEnabled:              req.statsEnabled,
+		AdditionalQueueDimensions: addlQueueDims,
+	}, nil
+}
+
+const ShouldQueryIngestersQueueDimension = "ingester"
+const ShouldQueryStoreGatewayQueueDimension = "store-gateway"
+const ShouldQueryIngestersAndStoreGatewayQueueDimension = "ingester-and-store-gateway"
+
+func (a *frontendToSchedulerAdapter) extractAdditionalQueueDimensions(
+	ctx context.Context, request *httpgrpc.HTTPRequest, now time.Time,
+) ([]string, error) {
+	var err error
+
+	httpRequest, err := httpgrpc.ToHTTPRequest(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	tenantIDs, err := tenant.TenantIDs(httpRequest.Context())
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case querymiddleware.IsRangeQuery(httpRequest.URL.Path):
+		start, end, _, err := querymiddleware.DecodeRangeQueryTimeParams(httpRequest)
+		if err != nil {
+			return nil, err
+		}
+		return a.queryComponentQueueDimensionFromTimeParams(tenantIDs, start, end, now), nil
+	case querymiddleware.IsInstantQuery(httpRequest.URL.Path):
+		time, err := querymiddleware.DecodeInstantQueryTimeParams(httpRequest)
+		if err != nil {
+			return nil, err
+		}
+		return a.queryComponentQueueDimensionFromTimeParams(tenantIDs, time, time, now), nil
+	case querymiddleware.IsLabelsQuery(httpRequest.URL.Path):
+		start, end, err := querymiddleware.DecodeLabelsQueryTimeParams(httpRequest)
+		if err != nil {
+			return nil, err
+		}
+		return a.queryComponentQueueDimensionFromTimeParams(tenantIDs, start, end, now), nil
+	case querymiddleware.IsCardinalityQuery(httpRequest.URL.Path):
+		// cardinality only hits ingesters
+		return []string{ShouldQueryIngestersQueueDimension}, nil
+	default:
+		// no query time params to parse; cannot infer query component
+		return nil, nil
```

Comment:
Main open question here is: do we give this case no additional queue dimension or do we make the "ingester-and-store-gateway" the "default"/"we don't know" queue.

If we do not give it in additional queue dimensions it would get enqueued with only the tenant dimension, not into the subqueues. This works just fine as mentioned before - any requests from the v1 frontend will work this way as well.

But I see a case for doing our best to place everything coming in from the v2 frontend into one of the subqueues, and it seems appropriate to make `"ingester-and-store-gateway"` the default.

Reply:
What queries would end up in this bucket? 

If we're not expecting anything to end up in this bucket, then logging that we got a type of query we didn't understand and assigning them to the tenant-only dimension seems reasonable to me.

Reply:
added a warning log.
the only thing that would end up in this bucket is a something that's not recognized as a range, instant, labels, or cardinality query, but somehow passed all validation above this in the query-frontend stack. I do not know if such a query exists.